compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java

```diff
@@ -1360,22 +1363,81 @@ private boolean isValidAnnotationType(BType type) {
         }
 
         switch (type.tag) {
+            case TypeTags.MAP:
+                BType constraintType = ((BMapType) type).constraint;
+                return isAnyDataOrReadOnlyTypeSkippingObjectType(constraintType);
             case TypeTags.RECORD:
-//                BRecordType recordType = (BRecordType) type;
-//                return recordType.fields.stream().allMatch(field -> types.isAnydata(field.type)) &&
-//                        (recordType.sealed || types.isAnydata(recordType.restFieldType));
+                BRecordType recordType = (BRecordType) type;
+                for (BField field : recordType.fields.values()) {
+                    if (!isAnyDataOrReadOnlyTypeSkippingObjectType(field.type)) {
+                        return false;
+                    }
+                }
+
+                BType recordRestType = recordType.restFieldType;
+                if (recordRestType == null || recordRestType == symTable.noType) {
+                    return true;
+                }
+
+                return isAnyDataOrReadOnlyTypeSkippingObjectType(recordRestType);
+            case TypeTags.ARRAY:
+                return isValidAnnotationType(((BArrayType) type).eType);
+        }
+
+        return types.isAssignable(type, symTable.trueType);
+    }
+
+    private boolean isAnyDataOrReadOnlyTypeSkippingObjectType(BType type) {
```

Comment:
Shall we create an issue to track fixing std lib deviations and add back the object validation to annotations?

Reply:
Issue #24217

Reply:
Created an issue to do language changes after StandardLib changes: https://github.com/ballerina-platform/ballerina-lang/issues/24246